### PR TITLE
LPS-113561 Avoid Liferay.provide in document-library-web

### PR DESCRIPTION
**Pull request by julien**
As part of [LPS-113561](https://issues.liferay.com/browse/LPS-113561) we
eventually want to deprecate the `Liferay.provide` function

To test the change for the `<portlet:namspace />selectFileEntryType`
function:

- Navigate to "Content and Data > Documents and Media"
- Click on the "+" button and choose "Folder"
- Type a name for the folder you want to create
- Click on the "Save" button
- Click on the kebab menu to the right of the folder you created and
  choose "Edit"
- Select the radio button with the label
  "Define Specific Document Type Restrictions and Workflow for This
   Folder (A folder)"
- Click on "Select Document Type"
- In the modal window choose one of the document types and click on
  "Choose"
- Click on the "Save button"

As a result, a notification should be displayed letting you know your
request has completed successfully and no JS errors should appear in the
browser's console.

To test the change for the `<portlet:namspace />updateMultipleFiles`
function:

- Navigate to "Content and Data > Documents and Media"
- Click on the "+" button and choose "Multiple Files Upload"
- Click on the "Select Files" button
- Upload at least 2 files (images or text documents are OK)
- Click on the "Publish" button

As a result, the files you uploaded should be visible in the "Documents"
list and no JS errors should appear in the console.

### CODE REVIEW

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:test:relevant

File: modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp [hunk 1]
```diff
@@ -380,78 +380,76 @@ renderResponse.setTitle(headerTitle);
...
+	window['<portlet:namespace />selectFileEntryType'] = function (
+		fileEntryTypeId,
+		fileEntryTypeName
+	) {
+		var A = AUI();
 
-			var searchContainer = Liferay.SearchContainer.get(
-				'<portlet:namespace />dlFileEntryTypesSearchContainer'
-			);
+		var searchContainer = Liferay.SearchContainer.get(
```
Comment (jbalsas) [inline modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp:389]: I think we went through this in previous PRs... now you depend on `liferay-search-container` here implicitly and there's no way to know when this will work and when it won't.

File: modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp [hunk 2]
```diff
@@ -380,78 +380,76 @@ renderResponse.setTitle(headerTitle);
...
+					fileEntryTypeId
+				);
+			</c:otherwise>
+		</c:choose>
 
-			searchContainer.updateDataStore();
+		searchContainer.updateDataStore();
 
-			var select = A.one('#<portlet:namespace />defaultFileEntryTypeId');
+		var select = A.one('#<portlet:namespace />defaultFileEntryTypeId');
```
Comment (jbalsas) [inline modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp:431]: Also seem to depend on `aui-node` here without declaring it (since you removed the `liferay-search-container` dependency that would bring it transitively.

File: modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp (hunk 1 above)
Comment (julien) [reply modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp:389]: Yes, we did, but I still wanted to send this PR to talk about it:

I added the `use="liferay-search-container"` attribute to the `<aui:script>` tag [here](https://github.com/liferay-frontend/liferay-portal/blob/b60b4562efce413acaa6bf0c321e0620b8d1df69/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L333).

After deploying that, the call to the `savePage` function made [here](https://github.com/liferay-frontend/liferay-portal/blob/b60b4562efce413acaa6bf0c321e0620b8d1df69/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L82) failed saying that `<portlet:namespace />savePage` function was undefined. 
(The `savePage` function is declared [here](https://github.com/liferay-frontend/liferay-portal/blob/b60b4562efce413acaa6bf0c321e0620b8d1df69/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L366))

Comment (jbalsas) [reply modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp:389]: When you add `use` to `aui:script`, it will create a closure that will hide `function savePage` within it instead of pushing it to the global scope as it is without it.

Comment (julien) [reply modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp:389]: oh then, that makes sense - didn't see it
I'll fix that

**Comment by jbalsas** [comment]
Hey @julien, I see you pushed a commit and tagged me as a reviewer. Could you provide some context on to what you pushed and what I should be looking for?

What I'm seeing is we're still dealing with race conditions here, so I'd expect some analysis here like the one @wincent did for you yesterday at https://github.com/liferay-frontend/liferay-portal/pull/22#discussion_r441553819

Please, add those details so I know what to look and what did you already thought about.

Thanks!

**Comment by julien** [comment]
@jbalsas sure,

In my last commit, I added the `use` attribute with the `liferay-search-container` dependency because when I read the conversation we had [here](https://github.com/liferay-frontend/liferay-portal/pull/11#discussion_r440754124), I understood that it should be done to make the dependency on `liferay-search-container` explicit. (Did I misunderstand that?)

I also declared the `<portlet:namespace />openFileEntryTypeSelector`  and `<portlet:namespace />savePage` functions on `window` because as you told me, when using the `use` attribute with the `<aui:script>` tag, the code inside that tag is in a closure and I need those functions to be accessible from where they're called.

If I understand correctly, because of this change, we have a race condition: these functions could be called without having being defined and that would blow up (for example if we called those functions before the `liferay-search-container dependency` has loaded).

I'm going to analyse and try to provide more information.

**Comment by julien** [comment]
What I'm seeing here is that:

The `liferay-search-container` component [depends](https://github.com/liferay/liferay-portal/blob/0a1ebcca26d3eabf9e443f7aa371b20d9f85ccd5/modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/search_container.js#L336) on `aui-base` and `aui-component` which are available before these functions are called (probably not useful information in this case but I still wanted to put all the things I thought about so you can correct me if I'm wrong)

In the page in which these functions (`<portlet:namespace />openFileEntryTypeSelector`, `<portlet:namespace />saveForm` and `<portlet:namespace />selectFileEntryType`) are defined, there's also a [`<liferay-ui:search-container>`](https://github.com/liferay-frontend/liferay-portal/blob/ab8037adc78fcb70d0b40091709338c5bbbedfea/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L158) tag - I'm not sure at all about this but I see that we have `<aui:script use="liferay-search-container">` [here](https://github.com/liferay/liferay-portal/blob/e3ffac158e0ec5acc5c67069fbd7ba688d3c78d4/portal-web/docroot/html/taglib/ui/search_iterator/list.jsp#L357) and more or less the same thing [here](https://github.com/liferay/liferay-portal/blob/51a7047479aaaf5da6ceee9c00b864125e9b8b3c/portal-web/docroot/html/taglib/ui/search_iterator/lexicon/bottom.jspf#L34-L52), as I understand it, these JSP files are part of the `<liferay-ui:search-container>` tag (is that correct?) and which makes me wonder if the `liferay-search-container` module will be loaded when those functions are called.

In any case, since we do have a dependency on `liferay-search-container` and that we want to make it explicit, but we still want these functions to be defined eagerly, what could we do here?

**Comment by wincent** [comment]
Here's how I see it.

Before this change, in `edit_folder.jsp`, we had:

* Script tag, no closure, no declared dependencies, defines two (namespaced) global functions and one variable : `openFileEntryTypeSelector()`, `savePage()`, and `documentTypesChanged`.
* The first of those two functions uses `Liferay.SearchContainer.get()`, so it has an implicit dependency on the `liferay-search-container` AUI module.
* We use `Liferay.provide` to synchronously define a (namespaced) global function, `selectFileEntryType()`. This one also uses `Liferay.SearchContainer.get()`, but unlike the other one, declares the `liferay-search-container` dependency explicitly.

After this change:

* Use of `use` to depend on `liferay-search-container` means that it won't be eagerly evaluated. That means `openFileEntryTypeSelector()`, `savePage()` and `documentTypesChanged` won't be set on `window` until the dependency is loaded. In fact, `documentTypesChanged` won't be set on `window` at all because it is in a closure now.
* In the absence of `Liferay.provide`, we now stick `selectFileEntryType` on `window`,  but note that the use of `use` means that is is no longer happening eagerly (we wait for the `liferay-search-container` dependency). We're using `aui-node` methods like `A.one` without declaring our dependency on them, but we get them transitively due to the `use` of `liferay-search-container`.

That's just `edit_folder.jsp`. To stop this comment from growing too big, I'll stop there, but the above changes would require the following questions to be answered. (And note, it may be that this all works as-is, but in order to avoid risking a regression, it's a good exercise to go through and ask the questions.)

* Who are the callers of `openFileEntryTypeSelector()`, `savePage()` and `selectFileEntryType()`?
  * At the time those callers can run, how do we know that `liferay-search-container` has been loaded, and that the functions are therefore actually defined?
* Who reads the (formerly) global `documentTypesChanged`? (ie. do we need to make that a global again?)
* How early does `aui-node` load in the page life cycle? (ie. look in [AUITopHeadResources](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-aui-web/src/main/java/com/liferay/frontend/js/aui/web/internal/servlet/AUITopHeadResources.java) to see whether it is "core", which means loaded very early no matter what, or "preloaded", which means that it may or may not be loaded early based on user settings — but note that in any case loading is still asynchronous, so there may be a race before `A.one` etc are available... unlikely as it may seem that this would ever be a problem, it is still probably worthwhile migrating off of `A.one`).

Of all of these, for me the most important question is the first one, because withouth that, the analysis can't really proceed. As I said, stopping here because I don't want this to spiral out of control, but I'd be applying exactly the same kind of analysis to upload_multiple_file_entries.jsp.

**Comment by julien** [comment]
@wincent, here's my reply to those questions

> Who are the callers of `openFileEntryTypeSelector()`, `savePage()` and `selectFileEntryType()`? 

Here are the respective callers of these functions.

For `openFileEntryTypeSelector()`,  it's [this](https://github.com/liferay/liferay-portal/blob/master/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L222)

for `savePage()` it's [this](https://github.com/liferay/liferay-portal/blob/master/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L82)

and for `selectFileEntryType()`, it's [this](https://github.com/liferay/liferay-portal/blob/master/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L353)

> At the time those callers can run, how do we know that liferay-search-container has been loaded, and that the functions are therefore actually defined?

Well, if I check `Liferay.SearchContainer` in those functions it's never `undefined ... I don't know if that seems like a valid answer

> Who reads the (formerly) global `documentTypesChanged`? (ie. do we need to make that a global again?) 

Here are the places when `documentTypesChanged` is read

https://github.com/liferay/liferay-portal/blob/master/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L334

https://github.com/liferay/liferay-portal/blob/master/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L372

https://github.com/liferay/liferay-portal/blob/master/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#414

https://github.com/liferay/liferay-portal/blob/master/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#494

I checked and tested (I know that's going to make your smile), and it doesn't' need to be made global again.

> How early does `aui-node` load in the page life cycle? (ie. look in `AUITopHeadResources` to see whether it is "core", which means loaded very early no matter what, or "preloaded", which means that it may or may not be loaded early based on user settings — but note that in any case loading is still asynchronous, so there may be a race before `A.one` etc are available... unlikely as it may seem that this would ever be a problem, it is still probably worthwhile migrating off of `A.one`).

I see `aui-node` in `_FILE_NAMES_AUI_PRELOAD` [here](https://git.io/JfFAB), so it's "preloaded" based on the user settings.

As whether it's worthwile migratting off of `A.node`: I personally think it's worth it, but I also think it would be great if the teams reponsible of this module could help.In the meantime, I'll do it in this PR, but we still have lots of AUI code in most modules.

File: modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp [hunk 3]
```diff
@@ -486,34 +490,42 @@ renderResponse.setTitle(headerTitle);
...
+			var selectContainer = document.getElementById(
+				'<portlet:namespace />restrictionTypeWorkflow'
 			);
 
-			var fileEntryTypesCount = select.get('children').size();
+			var fileEntryTypesCount = select.children.length;
 
 			if (fileEntryTypesCount == 0) {
-				selectContainer.hide();
+				selectContainer.classList.add('hide');
```
Comment (wincent) [inline modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp:513]: (No need to update this pull now, but for the future) there's an `==` that could be switched to `===` here, so keep your eyes open for things like that as you drive through older files. As you know, some of these will be autofixed when we roll out https://github.com/liferay/eslint-config-liferay/pull/171 (and this would be one of them), but here are others that will require manual review, so it's a good idea to fix stuff as we see it.

File: modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp [hunk 4]
```diff
@@ -486,34 +490,42 @@ renderResponse.setTitle(headerTitle);
...
-				restrictionTypeWorkflow.show();
+				restrictionTypeWorkflow.classList.remove('hide');
+				restrictionTypeWorkflow.removeAttribute('hidden');
+				restrictionTypeWorkflow.style.display = '';
 			}
 			else {
-				selectContainer.show();
+				selectContainer.classList.remove('hide');
+				selectContainer.removeAttribute('hidden');
+				selectContainer.style.display = '';
```
Comment (wincent) [inline modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp:528]: Given that this is a repetitive multi-step operation I would have declared a couple of little local helper functions inside the closure:

```js
function show(element) {
    element.classList.remove('hide');
    element.removeAttribute('hidden');
    element.style.display = '';
}

function hide(element) {
    element.classList.add('hide');
    element.setAttribute('hidden', 'hidden');
    element.style.display = 'none';
}
```

(No need to make this change; just pointing out a possible refactoring.)

Comment (julien) [reply modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp:528]: You're right, and when replacing AUI's `hide` and `show`, I've had to do this several times in various modules. 

I don't know if this should be in our "core" utils (which I didn't want to "pollute" until we spoke about it), but it's used quite a lot.

Comment (wincent) [reply modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp:528]: I know that for some people, keeping things "DRY" is a religion, and that means pushing stuff into a core module if at all possible. Personally, I wouldn't take that step because:

1. In order to be truly DRY, you have to use it everywhere, which means auditing the code base, figuring out how to import it everywhere it is needed (which may be complicated because some of the uses will be in JSPs files and you'll be back in that same territory where you've spent the last few weeks of analyzing dependency and ordering issues etc), then you need to educate people to use the new tool (write documentation, post an announcement, encourage use), and if you want to be sure that they use it you need to write a lint too that catches usages of the "wrong API" (may be nontrivial). If you extract but don't do all the rest, you'll just be contributing to [the "15 competing standards" problem](https://xkcd.com/927/), where we wind up with one more way of doing the same thing.
2. The logic is so simple that extracting it out is akin to making an [is-odd](https://www.npmjs.com/package/is-odd) package (half-a-million weekly downloads, BTW). Abstraction is good when it enables you to address more complex problems by making things more intellectually manageable. In this case, the abstraction does nothing of the sort; it just declutters things by eliminating some repeated trivial operations.
3. I don't think we should be in the business of making it too easy to manipulate the DOM from inside JSP files. On the contrary, we need to be investing in infrastructure improvements that make it easier to use better tools (eg. React). When we can't do that due to performance reasons or other concerns, it's our job as an infra team to work on eliminating the root causes of those problems until we get to the point where people can develop on Liferay with one simple mental model ("I just use React") in a systematic way, and we take care of the details.

Comment (jbalsas) [reply modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp:528]: You beat me to it, @wincent!

moral_support

**Comment by wincent** [comment]
> Well, if I check `Liferay.SearchContainer` in those functions it's never `undefined ... I don't know if that seems like a valid answer
>
> ...
>
> I checked and tested (I know that's going to make your smile), and it doesn't' need to be made global again.

I get the impression that the algorithm you're employing here is:

1. If "A" is loaded asynchronously, and "A" is needed by "B";
2. Run the code and check (ie. with a `console.log`, or a `debugger`, or just by running it in the browser and seeing it work) that when "B" needs "A", it has "A".

I think that's a useful check, but as I said in [this thread](https://github.com/liferay-frontend/liferay-portal/pull/22#discussion_r441505003):

> testing is great, but when it comes to _races_, seeing it work once (or twice, or 100 times) isn't proof that it is always going to work. The only way you can approach certainty about that is to analyze the data dependencies, which I think you might be getting at here.

and:

> testing is helpful, but when it comes to race conditions, testing alone doesn't tell you about whether something happens before something else or not (that's why they're called races). eg. you might have `A` and `B` getting defined in some indeterminate order. Like a coin flip, you might check 1, 2, 3, or N times and always see `A` first. But doesn't mean there isn't a race, just that you didn't see `B` win the race.

and:

> another manual test...
> 
> That's not the way to verify that we don't have a race bug.

Not all races are bad (obviously true, because if it weren't then all parallel computing would be busted) but many are. I think the Wikipedia page on [race conditions](https://en.wikipedia.org/wiki/Race_condition) does a decent job of explaining why then indeterminacy can be a source of bugs:

> A race condition arises in software when a computer program, to operate properly, depends on the sequence or timing of the program's processes or threads.
>
> A race condition can be difficult to reproduce and debug because the end result is nondeterministic and depends on the relative timing between interfering threads. Problems that occur when using systems can therefore disappear when running in debug mode, when additional logging is added, or when attaching a debugger, often referred to as a "Heisenbug". It is therefore better to avoid race conditions by careful software design.

So, in your comment above, I don't seen any explanation of whether there is a race, and if there is, why it is safe — I only see that you didn't see any problems in testing. I think there are three things to confirm:

1.  You got rid of the `aui-node` stuff, which is good.
2.  About the global you said:

    > I checked and tested

    If "checked" here means "tested" then that's not really the answer I was looking for, because in the presence of a race, checking/testing means basically nothing. On the other hand, if "checked" means "analyzed to establish that code reading from the variable *cannot* do so before the variable is initialized", then that's great. I would expect it to be something like this:

    > `documentTypesChanged` is set (to `false`) in [an `aui-script` tag](https://github.com/liferay/liferay-portal/blob/f96280593ff2831b732d873f4de92c484754968d/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L333) which has a dependency on `liferay-search-container`. In other words, it comes into existence only after `liferay-search-container` has loaded.
    >
    > Subsequently, there is one read and one write inside a couple of functions that are also defined inside that same `aui-script` tag. Because these functions can't possibly be called until defined, and they can't be defined until after `documentTypesChanged` has been initially set, those access are safe.
    >
    > The only remaining access (verified by `git grep documentTypesChanged`) is in a callback function defined inside [*another* `aui-script` tag](https://github.com/liferay/liferay-portal/blob/f96280593ff2831b732d873f4de92c484754968d/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L477). Because that tag *also* depends on `liferay-search-container`, and because it comes after the other `aui-script` tag (in the same file), we know that that access can only happen after `documentTypesChanged` is defined, so that one is safe too.

3.  About the other functions, your links make it look like all the callers are defined in the same file. That's good, because it simplifies the analysis. Taking the first one, I'd expect to see analysis like this:

    > [`openFileEntryTypeSelector()`](https://github.com/liferay/liferay-portal/blob/f96280593ff2831b732d873f4de92c484754968d/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L336) is defined inside an `aui-script` tag that depends on `liferay-search-container` (via `use`). This means that it won't be defined until the `liferay-search-container` dependency has been loaded.
    >
    > The sole caller (verified by `git grep openFileEntryTypeSelector`) is in the same file inside [a `<liferay-ui:icon>` tag](https://github.com/liferay/liferay-portal/blob/f96280593ff2831b732d873f4de92c484754968d/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L222).
    >
    > As such, there is a race here where a user could conceivably click the icon before the dependency has loaded and the function has been defined, which would lead to a JS error of the form `Uncaught ReferenceError: openFileEntryTypeSelector is not defined`.

I'm not going to analyze the others, but that's just to give you a flavor of the kind of thought process I'd expect you to go through when vetting a change like this. Obviously we're not going to write a Russian novel every time we remove a `Liferay.provide` call, but even if you don't write it all out, you *should* be doing the analysis IMO. I know we're going to pause work on this epic soon, but you can imagine that over time as we get better at analyzing this stuff, and we build up trust, your PR summary can become much briefer than what I wrote above.

In this specific instance, we still need to decide what to do about the race. I only looked at the first of the three functions, but the other two could easily be the same.

---

Separate topic: now that I look more closely at `documentTypesChanged` I think I see a bug, so I'm going to comment separately on that inline.

File: modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp [hunk 5]
```diff
@@ -488,34 +490,42 @@ renderResponse.setTitle(headerTitle);
...
-				'#<portlet:namespace />defaultFileEntryTypeId-' +
+			var option = document.getElementById(
+				'<portlet:namespace />defaultFileEntryTypeId-' +
 					link.getAttribute('data-rowId')
-			).remove();
+			);
+
+			option.parentElement.removeChild(option);
 
 			<portlet:namespace />documentTypesChanged = true;
```
Comment (wincent) [inline modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp:500]: This is the possible bug that I mentioned [here](https://github.com/liferay-frontend/liferay-portal/pull/11#issuecomment-647447036).

Before this PR, `<portlet:namespace />documentTypesChanged` was a global and now it is not, because we had:

```js
<aui:script>
    var <portlet:namespace />documentTypesChanged = false;
```

and we changed it to:

```js
<aui:script use="liferay-search-container">
    var <portlet:namespace />documentTypesChanged = false;
```

ie. it is now in an IIFE due to the `use` (horrible leaky abstraction, I know), which means that is is local to the IIFE.

But down here on line 496, this line is now broken. Note that this is in a *separate* `<aui:script use="liferay-search-container">` block, which means that it is its own closure. It cannot "see" the `<portlet:namespace />documentTypesChanged` in the other closure, and there is no `var` declaration here so this one *is* setting the variable on the window.

TL;DR: Both of these should be:

```js
window['<portlet:namespace />documentTypesChanged'] = ...;
```

Comment (wincent) [reply modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp:500]: Or, now that I look at them... just merge the two script tags together. It looks like they are two identical consecutive script tags:

```js
<aui:script use="liferay-search-container">
    // stuff
</aui:script>

<aui:script use="liferay-search-container">
    // more stuff
</aui:script>
```

They could just be:

```js<aui:script use="liferay-search-container">
    // stuff
    // more stuff
</aui:script>
```

I could have sworn I commented on a similar thing in a PR over the last few days, but it must have been another one... 🤔

Comment (wincent) [reply modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp:500]: Ah, it was [this one](https://github.com/wincent/liferay-portal/pull/344#pullrequestreview-429073581).

**Comment by julien** [comment]
@wincent, concerning the [`savePage`](https://github.com/liferay-frontend/liferay-portal/blob/fb0e289a2e9b4acd676abed243aa1dda799c1586/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L366-L381) function, it's also defined in the same [`<aui:script>`](https://github.com/liferay/liferay-portal/blob/fb0e289a2e9b4acd676abed243aa1dda799c1586/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L333) tag that depends on `liferay-search-container`, and is called (checked using `git grep savePage`) when this [`<aui:form>`](https://github.com/liferay/liferay-portal/blob/master/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L82) is submitted, so if the form is submitted before the `liferay-search-container` module is loaded it could throw an error 

The [`selectFileEntryType`](https://github.com/liferay-frontend/liferay-portal/blob/fb0e289a2e9b4acd676abed243aa1dda799c1586/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L383-L456) function is also defined in this (same) [`<aui:script>`](https://github.com/liferay/liferay-portal/blob/fb0e289a2e9b4acd676abed243aa1dda799c1586/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L333) tag that depends on `liferay-search-contrainer` and it's called from [here](https://github.com/liferay-frontend/liferay-portal/blob/fb0e289a2e9b4acd676abed243aa1dda799c1586/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L353-L356) (inside the `openFileEntryTypeSelector` function), which is when the user has selected something inside a modal window, since the caller is also defined in that same `<aui:script>` tag it seems less likely to fail.

**Comment by jbalsas** [comment]
Hey @julien, do you have updates on this one? If you don't, I can take it and you can focus on the other 2 PRs we have pending to see if we can wrap this up.

Let me know!

**Comment by julien** [comment]
@jbalsas I added a comment a bit earlier on with what I saw concerning the [`savePage`](https://github.com/liferay-frontend/liferay-portal/blob/fb0e289a2e9b4acd676abed243aa1dda799c1586/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L366-L381) and [`selectFileEntryType`](https://github.com/liferay-frontend/liferay-portal/blob/fb0e289a2e9b4acd676abed243aa1dda799c1586/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/edit_folder.jsp#L383-L456).

Concerning the [`updateMultipleFiles`](https://github.com/liferay-frontend/liferay-portal/blob/fb0e289a2e9b4acd676abed243aa1dda799c1586/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/upload_multiple_file_entries.jsp#L124) function, I know we could replace the calls to `A.one` and `A.all`, but if we really want to remove all AUI dependencies, we will have to replace/rewrite the [`A.LoadingMask`](https://github.com/liferay-frontend/liferay-portal/blob/fb0e289a2e9b4acd676abed243aa1dda799c1586/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/upload_multiple_file_entries.jsp#L157). 
The only thing I wanted to add about that function, is that it's called when [this](https://github.com/liferay-frontend/liferay-portal/blob/fb0e289a2e9b4acd676abed243aa1dda799c1586/modules/apps/document-library/document-library-web/src/main/resources/META-INF/resources/document_library/upload_multiple_file_entries_resources.jsp#L99) `<aui:form>` is submitted.

**Comment by jbalsas** [comment]
Ok, I'll take this one and push some changes.

**Comment by julien** [comment]
@jbalsas thanks

**Comment by jbalsas** [comment]
Hey @julien, I don't have edit access to your fork, so I can't push to this branch :(

Sending a new PR to https://github.com/liferay-frontend/liferay-portal/pull/50 😭
